### ISSUE TITLE
Handling Gateway reboot

### ISSUE THREAD

**Issue by neophob** [AlCalzone/node-tradfri-client#44]
I tested how my app behaves when the Tradfri Gateway reboots. First i though that I can use the `ping` function - but this won't work as it seems its using only the network layer (after the gateway reboots, ping returns true). 
I think the Tradfri Gateway stores the credentials used in the `connect` function only in memory - as after a reboot I cannot control the lamps anymore without issuing a new `connect` call.

How do you handle that case @AlCalzone ? One (high-level) idea might be that the library sends out a new event `gatewayError` - then the app that use this library can handle this case. 

More ideas about the `gatewayError` case:

- this event is fired when the library tries to send a command to the gateway - but no ack from the gateway is received (assuming that we use ack coap packages). 
- this event is also fired when there's a authentication issue (the app tries to use the credentials which were valid before the reboot of the gateway, but after the reboot they are not valid anymore)

What do you think about this? Or does the `error` event should already handle those cases?

**Comment by AlCalzone** [comment]
I‘m currently doing this check on the Application Level because I haven‘t found a fool-proof way to detect a hung-up gateway or similar issues. 
* Do regular pings (every 10s) as this is a very low-cost check. 
* When multiple successive pings fail (I use 3), call `reset()`
See also ioBroker.tradfri, src/main.ts, down at the bottom. 

I guess we could also use un-acked requests as an indicator, I‘ll think about it. 

To answer your other „question“:
The Connection is internally secured using DTLS, which is like SSL/TLS for UDP. The security params used are inherently temporary and only valid for one session (or a part thereof). So when one party reboots, the Connection has to be reestablished. 
That‘s why the `reset` method exists. It deletes everything related to the current DTLS session so another one is automatically created on the next request.

**Comment by neophob** [comment]
I hacked to together a similar solution on the application level. Its quite ugly at the moment but as soon as a ping failed, I define the Gateway as "faulty" - and as soon as the gateway is back I `reset` and reconnect to the gateway + observe the resources.

What I also figured out is, as soon as I disconnect the Ethernet connection I see some unhandled rejection:
```
UNHANDLED_REJECTION {"msg":"CoapClient was reset"}
UNHANDLED_REJECTION {"msg":"The DTLS handshake timed out"}
```
It might be my error, but I catch most of the failed promises. Any clue where they are thrown?

**Comment by AlCalzone** [comment]
Oh and btw, non-acked requests can also be detected by the corresponding Promises being rejected.

**Comment by neophob** [comment]
More new about the rejections:

```
Error: The DTLS handshake timed out
    at normalizeReason (/myproject/node_modules/node-coap-client/build/lib/DeferredPromise.js:5:16)
    at Promise.rej [as reject] (/myproject/node_modules/node-coap-client/build/lib/DeferredPromise.js:13:36)
    at Socket.onError (/myproject/node_modules/node-coap-client/build/CoapClient.js:964:25)
    at Object.onceWrapper (events.js:315:30)
    at emitOne (events.js:116:13)    
    at Socket.emit (events.js:211:7)
    at Socket.killConnection (/myproject/node_modules/node-dtls-client/build/dtls.js:236:22)
    at Socket.expectHandshake (/myproject/node_modules/node-dtls-client/build/dtls.js:152:22)
    at Timeout.<anonymous> (/myproject/node_modules/node-dtls-client/build/dtls.js:109:77)
    at ontimeout (timers.js:475:11)
    at tryOnTimeout (timers.js:310:5)
    at Timer.listOnTimeout (timers.js:270:5)"
```
looks like this rejection should be catched by this lib and resend via the error callback, what do you think?

```
Error: CoapClient was reset
    at normalizeReason (/myproject/node_modules/node-coap-client/build/lib/DeferredPromise.js:5:16)
    at Promise.rej [as reject] (/myproject/node_modules/node-coap-client/build/lib/DeferredPromise.js:13:36)
    at Function.reset (/myproject/node_modules/node-coap-client/build/CoapClient.js:191:33)
    at TradfriClient.reset (/myproject/node_modules/node-tradfri-client/build/tradfri-client.js:158:39)
    at tradfriClient.ping.catch.then (/myproject/app/lib/tradfriService.js:167:32)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
```

-> here it looks like the reset function can reject, according to the .ts file it returns void `reset(): void;`, doc error?

**Comment by AlCalzone** [comment]
> but as soon as a ping failed, I define the Gateway as "faulty" - and as soon as the gateway is back I reset and reconnect to the gateway + observe the resources.

To reiterate: As soon as the ping no longer fails, you reset the client? That shouldn't be necessary, because everything *should* be back to normal once the ping doesn't fail anymore. Resetting the client *can* bring things back to normality when the gateway seems dead but really isn't. A longer streak of failed pings *can* mean the gateway is actually not reachable.
See https://github.com/AlCalzone/ioBroker.tradfri/blob/master/src/main.ts#L618
I reset the client when the connection is dead for a while (in my case it sometimes just has a hiccup for a few seconds) and then start observing again when the connection is alive again.

Regarding the other unhandled promises:  
CoapClient reset should be handled indeed. Not sure about DTLS timeout, but I guess so too.

**Comment by AlCalzone** [comment]
I'm a bit hesitant to build in automatic reconnection and the like because I don't want to force users to wait 1.5 minutes (like I do). Some might reconnect more aggressively, some might have an unstable network..., you name it.

If you have a well-rounded plan on how to properly detect a connection loss, I'm open for suggestions. Maybe we should also start testing the connection on the network level, since the built-in ping function relies on the DTLS layer.

**Comment by AlCalzone** [comment]
> CoapClient was reset

Gets used as the rejection message for the promises for pending requests, observations and new connections when the CoAP client gets reset (by the user!) before those could be completed.
I want to notify the user that the promise cannot be fulfilled anymore, so this seems like a clean way. It should be documented though - basically anything you do can be rejected with this reason.

**Comment by neophob** [comment]
Thanks for splitting up the issues.

> To reiterate: As soon as the ping no longer fails, you reset the client? That shouldn't be necessary, because everything should be back to normal once the ping doesn't fail anymore. 

That didn't work when tested. I didn't implement a reconnect mechanism when I tested this. When the `ping` started to work again, I send commands to the gateway - but didn't get any feedback or could control the lamps. I did not send the observe command to the library, not sure if that's needed? But even if I don't get any feedback, controlling did not work (no error, no changed lamp state). I can retest if you thinks that *should* work.

> I'm a bit hesitant to build in automatic reconnection and the like because I don't want to force users to wait 1.5 minutes (like I do). Some might reconnect more aggressively, some might have an unstable network, maybe the gateway itself got stuck (happened a lot in earlier firmware versions)...

True story. Some ideas:

- reconnecting is hard to do right. it's also a common problem (in fact all implementations needs to handle it), maybe add an additional reconnect handler / set an option when initiating the library would be an option (so you can for example define the ping interval and failed count number)
- some apps might not handle the reconnection at all, this results either in "this lamps sucks" or "this lib sucks" (from a user view)

**Comment by AlCalzone** [comment]
> I send commands to the gateway - but didn't get any feedback or could control the lamps.

If the disconnection is due to the gateway restarting, this is expected. If it's the result of your network having a hiccup, things should go back to normal on their own. The method outlined in iobroker.tradfri is tested and works reliably.

**Comment by neophob** [comment]
Ok I did some more testing, here's the conclusion

- disconnect ethernet, then reconnect ethernet: if ping returns true, I'm able to send commands to tradfri without issues. only minor issue: if I try to reauthenticate while ethernet is disconnected, the `authenticate` function returns the error message `The security code is wrong` - which is wrong
- if i reboot the gatway `ping` will always return false - in this case I'm forced to call the `authenticate` function, then I call `clearObservers` and install all event handlers and observer devices and groups

for reference this is my current implementation:

```
reauthenticate(tradfri, securityCode) {
  return tradfri.authenticate(securityCode)
    .then((security) => {
      return tradfri.connect(security.identity, security.psk);
    });
}

  validateConnection(timeoutMs = 2000) {
    return this.tradfriClient.ping(timeoutMs)
      .catch((error) => {
        debug('ping failed', error.message);
        return false;
      })
      .then((pingResult) => {
        debug('pingResult', pingResult);
        if (this.reauthNeeded || pingResult === false) {
          this.gatewayReachable = false;
          this.gatewayOfflineCount++;
          if (this.reauthNeeded || this.gatewayOfflineCount === 5) {
            debug('try to reauthenticate', this.securityCode);
            this.reauthenticate(this.tradfriClient, this.securityCode)
              .then((result) => {
                debug('reauth result', result);
                this.gatewayOfflineCount = 0;
                this.reauthNeeded = false;
                this._startObserve();
              })
              .catch((error) => {
                debug('reauth failed', error.message);
                this.gatewayOfflineCount = 0;
                this.reauthNeeded = true;
              });
          }
        } else {
          this.gatewayReachable = true;
          this.gatewayOfflineCount = 0;
        }
      });
    }
}
```

this is WIP, some weak points:

- I should limit the reauthenticate calls, might be too expensive if the gateway is offline (or use an exponential increasing retry counter)

**Comment by AlCalzone** [comment]
I never had to reauthenticate on my gateway btw. The credentials generated from the first time survived all gateway reboots etc...

**Comment by neophob** [comment]
Thanks for the input, I will try that too. I could reproduce the behaviour you described, that after a gateway reboot ping didn't work anymore. So you say instead doing the reauth call I should use the reset function + re-observe resources?

**Comment by AlCalzone** [comment]
Exactly

**Comment by AlCalzone** [comment]
> if I try to reauthenticate while ethernet is disconnected, the authenticate function returns the error message The security code is wrong - which is wrong

fixed in 0.11.0 (warning: has breaking changes - see readme)

**Comment by neophob** [comment]
Oh nice, new releases!

I tried to use the reset call instead the reauth. code. that worked somehow - I can control my lamps but I don't get any device updates anymore. What i observed was, that the reauth function failed first (with a `unexpected response (4.05) to getPSK()` errror when i use the pretty aggressive 2s loop), but the reset might missed that somehow (or I made a mistake). anyway I'll try the PR with the reconnection now
